codec-http2/src/main/java/io/netty/handler/codec/http2/HpackHuffmanDecoder.java

```diff
@@ -4668,50 +4669,77 @@
 
+    private byte[] dest;
+    private int k;
+    private int state;
+    private int flags;
+
+    HpackHuffmanDecoder() { }
+
     /**
      * Decompresses the given Huffman coded string literal.
      *
      * @param buf the string literal to be decoded
      * @return the output stream for the compressed data
      * @throws Http2Exception EOS Decoded
      */
-    public static AsciiString decode(ByteBuf buf, int length) throws Http2Exception {
+    public AsciiString decode(ByteBuf buf, int length) throws Http2Exception {
         if (length == 0) {
             return AsciiString.EMPTY_STRING;
         }
-        byte[] dest = new byte[length * 8 / 5];
-        int k = 0;
-        int state = 0;
-        int flags = 0;
-        for (int i = 0; i < length; i++) {
-            byte input = buf.readByte();
-            int index = (state << 4) | ((input & 0xFF) >>> 4);
-            int row = HUFFS[index];
-            flags = row & 0x00FF00;
-            if ((flags & HUFFMAN_FAIL_SHIFT) != 0) {
-                throw BAD_ENCODING;
+        dest = new byte[length * 8 / 5];
+        try {
+            int readerIndex = buf.readerIndex();
+            // Using ByteProcessor to reduce bounds-checking and reference-count checking during byte-by-byte
+            // processing of the ByteBuf.
+            int endIndex = buf.forEachByte(readerIndex, length, this);
+            if (endIndex == -1) {
+                // We did consume the requested length
+                buf.readerIndex(readerIndex + length);
+                if ((flags & HUFFMAN_COMPLETE_SHIFT) != HUFFMAN_COMPLETE_SHIFT) {
+                    throw BAD_ENCODING;
+                }
+                return new AsciiString(dest, 0, k, false);
             }
-            if ((flags & HUFFMAN_EMIT_SYMBOL_SHIFT) != 0) {
-                dest[k++] = (byte) (row & 0xFF);
-            }
-            state = row >> 16;
 
-            index = (state << 4) | (input & 0x0F);
-            row = HUFFS[index];
-            flags = row & 0x00FF00;
-            if ((flags & HUFFMAN_FAIL_SHIFT) != 0) {
-                throw BAD_ENCODING;
-            }
-            if ((flags & HUFFMAN_EMIT_SYMBOL_SHIFT) != 0) {
-                dest[k++] = (byte) (row & 0xFF);
-            }
-            state = row >> 16;
-        }
-        if ((flags & HUFFMAN_COMPLETE_SHIFT) != HUFFMAN_COMPLETE_SHIFT) {
+            // The process(...) method returned before the requested length was requested. This means there
+            // was a bad encoding detected.
+            buf.readerIndex(endIndex);
             throw BAD_ENCODING;
+        } finally {
+            dest = null;
+            k = 0;
+            state = 0;
+            flags = 0;
         }
-        return new AsciiString(dest, 0, k, false);
     }
 
-    private HpackHuffmanDecoder() { }
+    /**
+     * <strong>This should never be called from anything but this class itself!</strong>
+     */
+    @Override
```

Comment:
You can add `@Deprecated` to make the IDE show it, and suppress the warning in `decode()`

Reply:
Actually this not works as the caller will be `ByteBuf` itself and not the user directly. So lets just keep the javadocs.